src/box/sql/build.c

```diff
@@ -2117,6 +2169,576 @@ sql_drop_table(struct Parse *parse_context)
+static const char *
+fk_name(struct Parse *parse)
+{
+	struct Token *name_token = &parse->create_fk_def.base.base.name;
+	if (name_token != NULL && name_token->z != NULL) {
+		char *str = sql_name_from_token(parse->db, name_token);
+		if (str == NULL)
+			return NULL;
+		const char *name = tt_sprintf("%s", str);
+		sqlDbFree(parse->db, str);
+		return name;
+	}
+	const char *space_name = parse->create_column_def.space->def->name;
+	return tt_sprintf("fk_unnamed_%s_%u", space_name, parse->constr_count);
+}
+
+/** Return space_id of the parent space in the new FOREIGN KEY constraint. */
+static int
+fk_parent_space_id(struct Parse *parse, uint32_t *space_id)
+{
+	struct Token *parent_token = parse->create_fk_def.parent_name;
+	char *parent_name = sql_name_from_token(parse->db, parent_token);
+	if (parent_name == NULL)
+		return -1;
+	const char *space_name;
+	if (parse->create_column_def.space == NULL) {
+		struct alter_entity_def *alter_def =
+			&parse->create_fk_def.base.base.base;
+		space_name = alter_def->entity_name->a[0].zName;
+	} else {
+		space_name = parse->create_column_def.space->def->name;
+	}
+	if (strcmp(parent_name, space_name) == 0) {
+		sqlDbFree(parse->db, parent_name);
+		*space_id = 0;
+		return 0;
+	}
+
+	struct space *parent_space = space_by_name(parent_name);
+	if (parent_space != NULL) {
+		sqlDbFree(parse->db, parent_name);
+		*space_id = parent_space->def->id;
+		return 0;
+	}
+	diag_set(ClientError, ER_NO_SUCH_SPACE, parent_name);
+	sqlDbFree(parse->db, parent_name);
+	parse->is_aborted = true;
+	return -1;
+}
+
+/** Create a new field FOREIGN KEY constraint. */
+static int
+field_fk_create(struct Parse *parse, struct tuple_constraint_def *cdef,
+		const char *name)
+{
+	uint32_t space_id;
+	if (fk_parent_space_id(parse, &space_id) != 0)
+		return -1;
+
+	struct ExprList *parent_cols = parse->create_fk_def.parent_cols;
+	assert(parent_cols != NULL);
+	if (parent_cols->nExpr != 1) {
+		diag_set(ClientError, ER_CREATE_FK_CONSTRAINT, name,
+			 "number of referencing and referencing columns for "
+			 "foreign key do not match");
+		parse->is_aborted = true;
+		return -1;
+	}
+
+	const char *field_name = parent_cols->a[0].zName;
+	if (field_name == NULL)
+		return -1;
+
+	uint32_t name_len = strlen(name);
+	uint32_t field_name_len = strlen(field_name);
+	uint32_t size = name_len + field_name_len + 2;
+	char *str = region_alloc(&parse->region, size);
+	if (str == NULL) {
+		diag_set(OutOfMemory, size, "region_alloc", "str");
+		parse->is_aborted = true;
+		return -1;
+	}
+	memcpy(str, name, name_len + 1);
+	cdef->name = str;
+	cdef->name_len = name_len;
+	cdef->type = CONSTR_FKEY;
+	cdef->fkey.space_id = space_id;
+	cdef->fkey.field_mapping_size = 0;
+	cdef->fkey.field.id = 0;
+	cdef->fkey.field.name_len = field_name_len;
+	str += name_len + 1;
+	memcpy(str, field_name, field_name_len + 1);
+	cdef->fkey.field.name = str;
+	return 0;
+}
+
+/** Add a new field FOREIGN KEY constraint to space format. */
```

Comment:
True, though `child_cols` can contain only one column. Still, in current version I create tuple fk when `child_cols` is not NULL and field fk otherwise.